### PR TITLE
[perf] feat: Add Aysnc Activation Offload

### PR DESCRIPTION
**Pull request by Zhang1Sheng**
### What does this PR do?

> Add async activation offload — a memory optimization that offloads hidden states from GPU to CPU during the forward pass and prefetches them back during the backward pass, using dedicated CUDA/NPU streams to overlap transfers with computation. Combined with gradient checkpointing (recomputation for intermediate activations), this enables training with significantly reduced peak GPU memory with minimal throughput impact.

Key design decisions:

- **Class-level `__call__` patching**: places `async_save_on_cpu` **outside** the `GradientCheckpointingLayer` checkpoint boundary so that hidden states saved by `_NoopSaveInputs.apply` pass through the offload hook, while intermediate activations are handled by GC recomputation — yielding zero persistent GPU memory for both.
- **Auto-discovery from `model._no_split_modules`**: mirrors the FSDP sharding pattern — offload targets are the same transformer block classes used for FSDP boundaries, requiring zero extra configuration.
- **Single-stream architecture with prefetch**: uses one swap stream for both D2H and H2D; during backward, launches H2D for layer N-1's tensors while computing layer N's backward, overlapping transfers with computation.

### Checklist Before Starting

- Search for relative PRs/issues and link here: ...
- PR title follows `[{modules}] {type}: {description}` format (enforced by [check_pr_title.yml](.github/workflows/check_pr_title.yml))
  - **Allowed modules:** `agent`, `ci`, `ckpt`, `config`, `data`, `dist`, `docker`, `docs`, `logging`, `lora`, `misc`, `model`, `omni`, `optim`, `ops`, `parallel`, `perf`, `release`, `task`, `trainer`
  - **Allowed types:** `feat`, `fix`, `refactor`, `chore`, `test`
  - Breaking changes: prepend `[BREAKING]` — e.g. `[BREAKING][parallel, model] feat: dynamic batching`

### Test

- [x] Unit tests in `tests/distributed/test_async_activation_offload.py`:
  - **Core components**: `SwapTensor` lifecycle (D2H/H2D state machine), `GetCnt` key generation (unique keys across blocks, multi-pass handling), `OffloadManager` singleton behavior, `async_save_on_cpu` pack/unpack round-trip correctness
  - **Module patching**: `get_offload_modules_from_class_names` auto-discovery, `async_offload_modules` per-instance attribute injection (`_veomni_offload_layer_idx`, `_veomni_offload_depth`), class-level patching idempotency
  - **Argument validation**: `enable_async_activation_offload` raises `ValueError` when gradient checkpointing is disabled
  - **`apply_async_activation_offload` end-to-end**: applies to models with `_no_split_modules`, warns for models without
  - **Integration**: model forward with patched `__call__` produces correct output/logits matching unpatched baseline

### Enable via config

```yaml
train:
  gradient_checkpointing:
    enable: true              # Required
  accelerator:
    offload_config:
      enable_async_activation_offload: true
```

#### Enable via CLI

```shell
bash train.sh tasks/train_text.py configs/text/qwen3-moe.yaml \
    --train.gradient_checkpointing.enable true \
    --train.accelerator.offload_config.enable_async_activation_offload true
```

#### Programmatic API

Called automatically by `build_parallelize_model` when the config flag is set:

```python
from veomni.distributed.async_offloading import apply_async_activation_offload

# Offload targets auto-derived from model._no_split_modules
apply_async_activation_offload(model)
```

#### Low-level context (for custom integration)

```python
from veomni.distributed.async_offloading import async_save_on_cpu

with async_save_on_cpu(block_idx=0, depth=12, prefetch=True):
    output = module(hidden_states)
```

### Design & Code Changes

| File | Change |
|------|--------|
| `veomni/distributed/async_offloading.py` | **New (498 lines).** Core implementation: `SwapTensor` (D2H/H2D lifecycle with pinned CPU buffer + event synchronization), `OffloadManager` (process-wide singleton registry with key-based storage and ref-counting), `GetCnt` (per-block tensor counting for unique offload keys, supports VLM multi-forward), `async_save_on_cpu` (`saved_tensors_hooks` context that intercepts and offloads hidden states), class-level `__call__` patching via `_patch_class_call`, `apply_async_activation_offload` entry point |
| `veomni/utils/device.py` | **New (+20 lines).** CUDA/NPU-agnostic stream/event utilities: `create_stream`, `create_event`, `get_current_stream`, `switch_to_specified_stream` |
| `veomni/arguments/arguments_types.py` | **New.** `enable_async_activation_offload: bool` field in `OffloadConfig`; `_validate_offload()` enforces `gradient_checkpointing.enable=True` at config-parse time |
| `veomni/distributed/torch_parallelize.py` | Hook `apply_async_activation_offload(model)` **after** GC enable but **before** FSDP2 sharding, gated on `enable_async_activation_offload` |
| `veomni/trainer/base.py` | Skip global fwd/bwd activation offload contexts when async offload is enabled (offloading is handled per-module via `saved_tensors_hooks`); use `nullcontext()` for both |
| `docs/key_features/async_activate_offload.md` | **New (245 lines).** Full documentation: overview, quick start, motivation, stream-based overlap, GC interaction, auto-discovery, configuration reference, core API, implementation details (SwapTensor lifecycle, OffloadManager singleton, class-level patching, prefetch mechanism), limitations |
| `tests/distributed/test_async_activation_offload.py` | **New (121 lines).** Unit tests with `ToyModel`/`ToyDecoderLayer` — core components, module patching, argument validation, end-to-end forward pass correctness |

### Architecture

```
Forward pass:
  Default Stream: [Compute Layer 0] [Compute Layer 1] [Compute Layer 2] ...
  Swap Stream:    [D2H Layer 0]     [D2H Layer 1]     [D2H Layer 2]    ...
                       ↑ overlapped with compute

Backward pass:
  Default Stream: [Backward Layer N] [Backward Layer N-1] ...
  Swap Stream:    [H2D Layer N-1]    [H2D Layer N-2]
                       ↑ prefetch overlaps with backward compute

saved_tensors_hooks stack order:
  [async_save_on_cpu (outer)]   ← intercepts hidden_states → offloaded to CPU
  [_checkpoint_hook (inner)]    ← handles intermediate activations → recomputed

SwapTensor state machine:
  device ──(D2H)──> host ──(H2D)──> device
```

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- Applied pre-commit checks
- Added/updated documentation
- If `tasks/` training scripts were moved or renamed: updated `docs/` examples and verified `python3 scripts/ci/check_doc_task_paths.py` passes (also enforced by the **Check doc task paths** CI workflow)
- Added tests to CI workflow (or explained why not feasible)

### CODE REVIEW

**Comment by Coach257** [comment]
#965

**Comment by YLlllllllllll** [comment]
This PR is superseded by `#965` (`feat/async-activation-offload`).

`#965` keeps the MindSpeed-style async activation offload idea, but it is the version that passed the device-API checker, auto-discovers `_no_split_modules`, avoids global class mutation, and has CI unit tests. Please close this one after `#965` merges.

cc @Luosuu @Coach257
